File: fs/before_checksum/956__number-of-music-playlists/cpp/number-of-music-playlists_benchmark_v2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <array>

class Solution {
public:
    int numMusicPlaylists(int N, int L, int K) {
        const int M = 1000000007;
        if (L < N) return 0;
        std::vector<std::vector<long long>> dp(2, std::vector<long long>(L + 1, 0));
        dp[0][0] = 1;
        for (int n = 1; n <= N; ++n) {
            int cur = n & 1;
            int prev = (n - 1) & 1;
            std::fill(dp[cur].begin(), dp[cur].end(), 0LL);
            dp[cur][n] = (dp[prev][n - 1] * n) % M;
            for (int l = n + 1; l <= L; ++l) {
                long long add_new = (dp[prev][l - 1] * n) % M;
                int replay_choices = std::max(n - K, 0);
                long long extend = (dp[cur][l - 1] * replay_choices) % M;
                dp[cur][l] = (extend + add_new) % M;
            }
        }
        return static_cast<int>(dp[N & 1][L] % M);
    }
};

int main() {
    std::vector<std::array<int, 3>> tests = {
        std::array<int,3>{1, 1, 0},
        std::array<int,3>{2, 2, 1},
        std::array<int,3>{2, 3, 0},
        std::array<int,3>{3, 3, 2},
        std::array<int,3>{3, 4, 1},
        std::array<int,3>{3, 5, 1},
        std::array<int,3>{4, 4, 3},
        std::array<int,3>{4, 6, 2},
        std::array<int,3>{5, 7, 0},
        std::array<int,3>{6, 8, 1}
    };

    Solution sol;
    volatile int sink = 0;

    const int iterations = 1;
    for (int iter = 0; iter < iterations; ++iter) {
        int checksum = 0;
        for (const auto& t : tests) {
            checksum += sol.numMusicPlaylists(t[0], t[1], t[2]);
        }
        sink += checksum;
        

    } std::cout << sink << std::endl;

    return 0;
}

File: fs/before_energy/956__number-of-music-playlists/cpp/number-of-music-playlists_benchmark_v2.py
# Time:  O(n * l)
# Space: O(l)

class Solution(object):
    def numMusicPlaylists(self, N, L, K):
        """
        :type N: int
        :type L: int
        :type K: int
        :rtype: int
        """
        M = 10**9+7
        dp = [[0 for _ in xrange(1+L)] for _ in xrange(2)]
        dp[0][0] = dp[1][1] = 1
        for n in xrange(1, N+1):
            dp[n % 2][n] = (dp[(n-1) % 2][n-1] * n) % M
            for l in xrange(n+1, L+1):
                dp[n % 2][l] = ((dp[n % 2][l-1] * max(n-K, 0)) % M + \
                                (dp[(n-1) % 2][l-1] * n) % M) % M
        return dp[N % 2][L]


if __name__ == "__main__":
    tests = [
        (1, 1, 0),
        (2, 2, 1),
        (2, 3, 0),
        (3, 3, 2),
        (3, 4, 1),
        (3, 5, 1),
        (4, 4, 3),
        (4, 6, 2),
        (5, 7, 0),
        (6, 8, 1)
    ]

    sol = Solution()
    sink = 0

    iterations = 1000
    for i in xrange(iterations):
        #checksum = 0
        for t in tests:
            r = sol.numMusicPlaylists(t[0], t[1], t[2])
        #sink = checksum

    # print sink